src/ripple/rpc/handlers/GetAggregatePrice.cpp

```diff
@@ -0,0 +1,321 @@
+
+    auto getField = [&](Json::StaticString const& field,
+                        unsigned int def =
+                            0) -> std::variant<std::uint32_t, error_code_i> {
+        if (params.isMember(field))
+        {
+            if (!params[field].isConvertibleTo(Json::ValueType::uintValue))
+                return rpcORACLE_MALFORMED;
+            return params[field].asUInt();
+        }
+        return def;
+    };
+
+    auto const trim = getField(jss::trim);
+    if (std::holds_alternative<error_code_i>(trim))
+    {
+        RPC::inject_error(std::get<error_code_i>(trim), result);
+        return result;
+    }
+
+    auto const timeThreshold = getField(jss::time_threshold, 0);
+    if (std::holds_alternative<error_code_i>(timeThreshold))
+    {
+        RPC::inject_error(std::get<error_code_i>(timeThreshold), result);
+        return result;
+    }
+
+    auto const& baseAsset = params[jss::base_asset];
+    auto const& quoteAsset = params[jss::quote_asset];
+
+    // Collect the dataset into bimap keyed by lastUpdateTime and
+    // STAmount (Number is int64 and price is uint64)
+    Prices prices;
+    for (auto const& oracle : params[jss::oracles])
+    {
+        if (!oracle.isMember(jss::oracle_sequence) ||
+            !oracle.isMember(jss::account))
+        {
+            RPC::inject_error(rpcORACLE_MALFORMED, result);
+            return result;
+        }
+        auto const sequence = oracle[jss::oracle_sequence].isConvertibleTo(
+                                  Json::ValueType::uintValue)
+            ? std::make_optional(oracle[jss::oracle_sequence].asUInt())
+            : std::nullopt;
+        auto const account =
+            parseBase58<AccountID>(oracle[jss::account].asString());
+        if (!account || account->isZero() || !sequence)
+        {
+            RPC::inject_error(rpcINVALID_PARAMS, result);
+            return result;
+        }
+
+        auto const sle = ledger->read(keylet::oracle(*account, *sequence));
+        iteratePriceData(context, sle, [&](STObject const& node) {
+            auto const& series = node.getFieldArray(sfPriceDataSeries);
+            // find the token pair entry with the price
+            if (auto iter = std::find_if(
+                    series.begin(),
+                    series.end(),
+                    [&](STObject const& o) -> bool {
+                        return o.getFieldCurrency(sfBaseAsset).getText() ==
+                            baseAsset &&
+                            o.getFieldCurrency(sfQuoteAsset).getText() ==
+                            quoteAsset &&
+                            o.isFieldPresent(sfAssetPrice);
+                    });
+                iter != series.end())
+            {
+                auto const price = iter->getFieldU64(sfAssetPrice);
+                auto const scale = iter->isFieldPresent(sfScale)
+                    ? -static_cast<int>(iter->getFieldU8(sfScale))
```

Comment:
Why do we need a negative sign for the `scale` variable? I suspect it has something to do with the internal mechanics of the `STAmount::mIsNegative` class, but I couldn't figure it out.

1. If I want to use negative values in the `PriceDataSeries` array, is that possible? (negative temperatures, depreciating yields, etc). We are using `unsigned` types internally, is this possible?

I modified an existing test-case to include negative values. The numbers are over-flowing because we're using `std::uint64_t` type. 

2. How can we represent fractional values? Ex: `{"XRP", "USD", "0.20", 1}` ? Do I need to do: `{"XRP", "USD", "2", 10}` i.e. Scale up the base asset by 10 to ensure a whole number ?

Reply:
We are converting the `AssetPrice` to `STAmount`. The `AssetPrice` is represented by `exponent` and `mantissa`. In our case the `mantissa`; i.e., the `Scale` is always negative. This is why we negate it. So it's not about the negative price value, it's about the negative `mantissa` to represent the fractional values.

1. Can't use the negative values
2. Correct

Reply:
Okay. Why are we using `originalPrice` for calculating statistics (`GetAggregatePrice`) ? I was of the understanding that, `SymbolPrice` represents the complete information about the `Asset`, because `SymbolPrice` already incorporates the `Scale` information. Is that not correct?

Reply:
`SymbolPrice` is mantissa and `Scale` is exponent, which if not 0 then it's always implied negative. We need both mantissa and exponent to fine the original price.

Reply:
okay, I understand now. thanks for explaining it